filterprotobuffer.go

```diff
@@ -0,0 +1,55 @@
+package dapper
+
+import (
+	"bytes"
+	"github.com/dogmatiq/iago/indent"
+	"github.com/dogmatiq/iago/must"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	"io"
+	"reflect"
+)
+
+// protobufMessageType is the reflect.Type for proto.Message
+var protobufMessageType = reflect.TypeOf((*proto.Message)(nil)).Elem()
+
+// ProtobufferFilter is a filter for proto.Message types. It shows the field names as defined in the .proto file
+// and hides implementation-specific internal state.
+func ProtobufferFilter(
+	w io.Writer,
+	v Value,
+	c Config,
+	p FilterPrinter,
+) (err error) {
+	if !v.DynamicType.Implements(protobufMessageType) {
+		return
+	}
+
+	must.WriteString(w, p.FormatTypeName(v))
+	must.WriteByte(w, '{')
+	defer must.WriteByte(w, '}')
+
+	message := v.Value.Interface().(proto.Message)
+
+	// prototext.MarshalOptions only allows tabs and spaces to be used for indentation.
+	// Fall back to spaces if the configured one is not valid, so we can still render something.
+	protoIndent := "    "
+	if len(bytes.Trim(c.Indent, " \t")) == 0 {
+		protoIndent = string(c.Indent)
+	}
+
+	formattedMessage := prototext.MarshalOptions{
+		Multiline:    true,
+		Indent:       protoIndent,
+		AllowPartial: true,
+		EmitUnknown:  true,
+	}.Format(message)
+
+	if len(formattedMessage) > 0 {
+		must.WriteByte(w, '\n')
+	}
+
+	must.WriteString(indent.NewIndenter(w, c.Indent), formattedMessage)
```

Comment:
Is there some known condition where the result of `Format()` can be empty?

We might want to do something different on line 52 in that case, such as rendering the "zero marker" to show that it's a zero-value, if that's why this happens.

Reply:
An empty struct, or a struct with all values set to its type zero-value.
I was rendering that zero marker initially, but then copied the behaviour from structs.